### PR TITLE
Fix `redundant_pattern` false positive

### LINKED ISSUE #4428: False-positive in `redundant_pattern` lint.
The nightly `slice_patterns` feature (see https://doc.rust-lang.org/nightly/unstable-book/language-features/slice-patterns.html) introduces the syntax `..` which matches any leftover elements in the slice, versus the `_` for a single item in stable Rust.

Clippy's `redundant_pattern` lint checks for `variable @ _` and correctly recommends it be changed to plain `variable`. However, there is (currently) no shorter syntax for `variable @ ..` in patterns, and the recommendation to change it to `variable` is incorrect and changes the meaning. There is also a cosmetic error in that it gives the binding as `_` rather than `..`.

I am using the latest nightly clippy ("clippy 0.0.212 (cd3df6be 2019-08-20)") and this is one example of the cascade of warnings it gives when choking on my parser library:
```
warning: the `stream @ _` pattern can be written as just `stream`
   --> src/bencode.rs:321:20
    |
321 |             [b'l', stream @ ..] => decode_list(stream).map(wrap),
    |                    ^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_pattern
```
(In case it's not obvious, this is within a `match { ... }` block.)

The desired behaviour is for Clippy is to not treat `..` as equivalent to `_` in patterns, because it isn't. Perhaps it was at some point in the evolution of this feature, but it isn't now.

### PR DESCRIPTION
**Pull request by JohnTitor**
Fix #4428

changelog: Fix `redundant_pattern` false positive

### CODE REVIEW

File: clippy_lints/src/misc_early.rs [hunk 1]
```diff
@@ -286,6 +309,20 @@ impl EarlyLintPass for MiscEarlyLints {
...
         }
+
+        if let PatKind::Ident(_, ident, Some(ref right)) = pat.node {
+            if let PatKind::Wild = right.node {
+                span_lint(
+                    cx,
+                    REDUNDANT_PATTERN,
+                    pat.span,
+                    &format!(
+                        "the `{} @ _` pattern can be written as just `{}`",
```
Comment (flip1995) [inline clippy_lints/src/misc_early.rs]: This seems to be a low hanging fruit to make an auto fixable lint out of this.

Would you like to do this in this PR?

Comment (JohnTitor) [reply clippy_lints/src/misc_early.rs]: I'd like to do that but I don't know how we do (replace with `span_lint_and_sugg`?).

Comment (flip1995) [reply clippy_lints/src/misc_early.rs]: Yes exactly. With this lint, this should be relatively self-explanatory. Make sure to read the suggestion rules of rustc here: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.Diagnostic.html#method.span_suggestion

Comment (JohnTitor) [reply clippy_lints/src/misc_early.rs]: applied suggested changes

File: tests/ui/patterns.rs [hunk 2]
```diff
@@ -1,8 +1,10 @@
 #![allow(unused)]
```
Comment (flip1995) [inline tests/ui/patterns.rs]:
```suggestion
// run-rustfix
#![allow(unused)]
```

Comment (flip1995) [reply tests/ui/patterns.rs]: Don't forget to run `cargo uitest` and `update-all-references.sh` after that

Comment (JohnTitor) [reply tests/ui/patterns.rs]: Okay, added the flag and `.fixed` file.

**Comment by flip1995** [comment]
Thanks!

@bors r+

**Pull request merged**